File: 206. Reverse Linked List/python/reverse_linked_list_test.py
import unittest

from reverse_linked_list import (
    Solution,
    ListNode,
)

RESULTS = (
	[5, 4, 3, 2, 1],
	[2, 1],
	[],
)

CASES = (
    [1,2,3,4,5],
    [1,2],
    [],
)

class TestSolution(unittest.TestCase):
    def runTest(self):
        a = Solution()
        b = a.reverseList
        d = a.reverseListS
        for i, v in enumerate(CASES):
            bl = []
            dl = []
            fl = ListNode() if v else None
            ll = fl
            for j in range(len(v)):
                ll.val = v[j]
                if j != len(v) - 1:
                    ll.next = ListNode()
                    ll = ll.next
            nb = b(fl)
            nd = d(fl)
            while not nb is None:
                bl.append(nb.val)
                nb = nb.next
            while not nd is None:
                dl.append(nd.val)
                nd = nd.next
            self.assertEqual(bl, RESULTS[i])
            self.assertEqual(dl, RESULTS[i])

File: 206. Reverse Linked List/go/reverse_linked_list_test.go
package reverseLinkedList

import (
	"reflect"
	"testing"
)

var Cases = [][]int{
	{1, 2, 3, 4, 5},
	{1, 2},
	{},
}
var Results = [][]int{
	{5, 4, 3, 2, 1},
	{2, 1},
	{},
}

func SimpleListToLL(a *[]int) *ListNode {
	ca := *a
	var fl *ListNode
	ll := &ListNode{}
	if len(ca) > 0 {
		fl = ll
	}
	for v := range ca {
		ll.Val = ca[v]
		if v != len(ca)-1 {
			ll.Next = &ListNode{}
			ll = ll.Next
		}
	}
	return fl
}

func LLToSimpleList(ll *ListNode) *[]int {
	dl := []int{}
	nl := ll
	for nl != nil {
		dl = append(dl, nl.Val)
		nl = nl.Next
	}
	return &dl
}

func TestSolution(t *testing.T) {
	for i := range Cases {
		fl := SimpleListToLL(&Cases[i])
		nl := reverseList(fl)
		dl := *LLToSimpleList(nl)
		// test
		want := Results[i]
		if !reflect.DeepEqual(dl, want) {
			t.Errorf("got %v, wanted %v", dl, want)
		}
	}
}

func TestSolutionOne(t *testing.T) {
	for i := range Cases {
		fl := SimpleListToLL(&Cases[i])
		nl := reverseListOne(fl)
		dl := *LLToSimpleList(nl)
		// test
		want := Results[i]
		if !reflect.DeepEqual(dl, want) {
			t.Errorf("got %v, wanted %v", dl, want)
		}
	}
}
